webrender/src/device/gfx/descriptor_set.rs

```diff
@@ -0,0 +1,167 @@
+    // TransformPalette sampler
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(3, DT::Sampler, SSF::ALL, false),
+];
+
+pub(super) const CLIP_SET_2: &'static [DescriptorSetLayoutBinding] = DEFAULT_SET_2;
+
+pub(super) const PRIMITIVE_SET_0: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding!(0, DT::SampledImage, SSF::ALL, false),
+    // RenderTasks
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(1, DT::SampledImage, SSF::ALL, false),
+    // GpuCache
+    descriptor_set_layout_binding!(2, DT::SampledImage, SSF::ALL, false),
+    // TransformPalette
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(3, DT::SampledImage, SSF::ALL, false),
+    // PrimitiveHeadersF
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(4, DT::SampledImage, SSF::ALL, false),
+    // PrimitiveHeadersI
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(5, DT::SampledImage, SSF::ALL, false),
+];
+
+pub(super) const PRIMITIVE_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0 sampler
+    descriptor_set_layout_binding!(12, DT::Sampler, SSF::ALL, false),
+    // Color1 sampler
+    descriptor_set_layout_binding!(13, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding!(14, DT::Sampler, SSF::ALL, false),
+    // PrevPassAlpha sampler
+    descriptor_set_layout_binding!(15, DT::Sampler, SSF::ALL, false),
+    // PrevPassColor sampler
+    descriptor_set_layout_binding!(16, DT::Sampler, SSF::ALL, false),
+    // Dither sampler
+    descriptor_set_layout_binding!(0, DT::Sampler, SSF::ALL, false),
+    // RenderTasks sampler
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(1, DT::Sampler, SSF::ALL, false),
+    // GpuCache sampler
+    descriptor_set_layout_binding!(2, DT::Sampler, SSF::ALL, false),
+    // TransformPalette sampler
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(3, DT::Sampler, SSF::ALL, false),
+    // PrimitiveHeadersF
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(4, DT::Sampler, SSF::ALL, false),
+    // PrimitiveHeadersI
+    // TODO: ShaderStageFlags should be VERTEX only here
+    descriptor_set_layout_binding!(5, DT::Sampler, SSF::ALL, false),
+];
+
+pub(super) const PRIMITIVE_SET_2: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0
+    descriptor_set_layout_binding!(0, DT::SampledImage, SSF::ALL, false),
+    // Color1
+    descriptor_set_layout_binding!(1, DT::SampledImage, SSF::ALL, false),
+    // Color2
+    descriptor_set_layout_binding!(2, DT::SampledImage, SSF::ALL, false),
+    // PrevPassAlpha
+    descriptor_set_layout_binding!(3, DT::SampledImage, SSF::ALL, false),
```

Comment:
these could go into a separate descriptor set, given that they only change per render target